### PR TITLE
Suffix arrays, needed for nbest processing

### PR DESCRIPTION
**Pull request by pkufool**
See  https://github.com/k2-fsa/snowfall/issues/232 and https://github.com/k2-fsa/k2/pull/779

### CODE REVIEW

**Comment by pkufool** [comment]
@danpovey What should the mean and variance be when there is no best-matching-position (tightest interval is root interval),  zeros?

**Comment by danpovey** [comment]
It's OK to use the root interval itself.
If the root interval has zero stats (which should not happen), you can use zeros.

File: k2/csrc/CMakeLists.txt [hunk 1]
```diff
@@ -86,7 +88,7 @@ else()
 endif()
 
 # the target
-add_library(context ${context_srcs})
+add_library(context ${context_srcs} ${context_cc_srcs})
```
Comment (csukuangfj) [inline k2/csrc/CMakeLists.txt:91]:
```suggestion
add_library(context ${context_srcs})
```

File: k2/python/k2/utils.py [hunk 2]
```diff
@@ -656,3 +656,93 @@ def random_fsa_vec(min_num_fsas: int = 1,
...
+              [ [ hi, my, name, is, eos ], [ bye, my, name, is, eos ] ], ... ]
+
+        where the words would actually be represented as integers,
+        The eos symbol is required if this code is to work as intended
+        (otherwise this code will not be able to recognize when we have reached
+        the beginnings of sentences when comparing histories).
+        bos symbols are allowed but not required.
+
+      scores:
+        An one dim torch.tensor with scores.size() == tokens.NumElements(),
```
Comment (csukuangfj) [inline k2/python/k2/utils.py:696]:
```suggestion
        A one dim torch.tensor with scores.size() == tokens.NumElements(),
```

**Pull request merged**